pallets/xyk/src/lib.rs

```diff
@@ -604,21 +603,25 @@ impl<T: Config> AMM<T::AccountId, AssetId, AssetPair, Balance> for Pallet<T> {
 
 		let pair_account = Self::get_pair_id(assets);
 
-		let asset_in_total = T::Currency::free_balance(assets.asset_in, &pair_account);
-		let asset_out_total = T::Currency::free_balance(assets.asset_out, &pair_account);
+		let asset_in_reserve = T::Currency::free_balance(assets.asset_in, &pair_account);
+		let asset_out_reserve = T::Currency::free_balance(assets.asset_out, &pair_account);
 
-		ensure!(amount <= asset_in_total / MAX_IN_RATIO, Error::<T>::MaxInRatioExceeded);
+		ensure!(
+			amount <= asset_in_reserve / MAX_IN_RATIO,
+			Error::<T>::MaxInRatioExceeded
+		);
 
 		let transfer_fee = if discount {
 			Self::calculate_discounted_fee(amount)?
 		} else {
 			Self::calculate_fee(amount)?
 		};
 
-		let sale_price = hydra_dx_math::calculate_out_given_in(asset_in_total, asset_out_total, amount - transfer_fee)
-			.map_err(|_| Error::<T>::SellAssetAmountInvalid)?;
+		let sale_price =
+			hydra_dx_math::calculate_out_given_in(asset_in_reserve, asset_out_reserve, amount - transfer_fee)
```

Comment:
Let's fix this as well and use checked math for fee subtraction

Reply:
done